### ISSUE TITLE
Wrong values while converting the model from Keras to ONNX

### ISSUE THREAD

**Issue by Rathna21** [onnx/keras-onnx#53]
Hello,

I have built a model for Semantic segmentation of street scenes in Keras and wanted to convert it into ONNX as I have to work in Windows ML.

I could successfully convert the model to ONNX and when I load the model and evaluating it, no error but if I print the output values they are all 0.

This is the following script I used to convert the same:

**import keras2onnx
import onnx
from keras.models import load_model

model = load_model(filepath="unet_full_model.h5")
onnx_model = keras2onnx.convert_keras(model=model, channel_first_inputs= input_1)
onnx.save_model(onnx_model, "new_unet.onnx")
print("DOne")**

Could someone please help me on this?

**Comment by jiafatom** [comment]
@Rathna21 Can you share the keras model file please?

**Comment by Rathna21** [comment]
@jiafatom Thanks for the reply:)

Yea please find the attached Keras model.

https://drive.google.com/file/d/1y2TP-2mhDkgkCOJKBTvPdAE0hTPJuECd/view?usp=sharing

**Comment by jiafatom** [comment]
I see the output mismatches the keras result. Need some time to debug (possibly check each layer to compare the output, which is non-trivial). Just wonder if you can give me a sub model, for example, only containing the first Conv2D and BatchNormalization, and then I can quickly test it to see whether it is correct.

**Comment by Rathna21** [comment]
@jiafatom 

Thanks for the reply.

What do you mean by a sub model?

Should I create a new model and train and send or how could this be done?

And I have one more query?
With your keras2onnx script I am trying to covert yolo model in keras to onnx. 
But at the end there is a transpose layer which changes the shape of the output?

Is there a way i can eliminate this?
Please find the screenshot.
![Screen Shot 2019-04-03 at 8 02 18 PM](https://user-images.githubusercontent.com/34045027/55501783-719a8a00-564b-11e9-9894-dab62d2f6d2f.png)

**Comment by jiafatom** [comment]
Yes, can you train a new model with less layers (just include the first Conv and Batchnormalization) to see the output matches? This can help locate which layer contains issues, especially I want to see Batchnormalization. 

About outputshape, the onnx model should match your keras model output shape, this is why it has a transpose here. Are you talking about the keras model output shape is different?

**Comment by Rathna21** [comment]
@jiafatom  Okay I will train the model with few layers and upload the keras model.


And about the output shape, I have converted **Darknet yolo model** into Keras using ya2dk script and want to convert the keras model to ONNX and work in Windows ML . 
The actual problem is, in Keras the way it accepts input and output shape is { batch_size, height , width, channels } but in Windows ML it is { batch_size , channels , height, width } 

So, i used your script and gave channels_first_input as my input so now the input shape in onnx model is like { batch_size, channels, height, width} But i want this channels first in output shape as well. 

Please find the attached issue. 
https://github.com/Microsoft/Windows-Machine-Learning/issues/205

I have created the above mentioned issue for some other problem and some one helped me with channels_first_input . But is there a way I can get channels_first for output as well?

ALSO: 

While converting my model which I have already sent you , I converted into onnx giving channels_first_input so, is this the problem for mismatch in output values?

Please do help me with this?

**Comment by jiafatom** [comment]
@Rathna21 the output shape depends on the keras model. I checked the model you shared with me, the input data shape is (2, 256, 512, 3), and the output shape is (2, 256, 512, 4) when I do model.predict(data) using keras. This is how keras gives you, and our onnx output shape matches this. 

channels_first_input is to give input shape channel first, which is the way works in WindowsML. We don't do channel_first_output, I don't see a reason why we do this, onnx output shape just match keras output shape. keras result is not channel_first_output, don't understand why you need channel_first_output?

(if you want to do channel_first_output, you can add a transpose layer in keras layer)

**Comment by Rathna21** [comment]
@jiafatom Correct as you said the model I shared with you has output shape (2,256,512,4). That is what even I want. Thats fine. The problem in that is my output values .


But I have another issue. I am working with yolo as well. I need to integrate the yolo model in Windows ML. YOLO in darknet is converted into keras using yad2k script. So, in keras it changes the input and output shape as (batch_size, height, width, channels) . But in windows ML its different. When I see yolo model in onnx model zoo, the output shape is (batch_size, channels , height, width) . Its the same in CORE ML as well. So, further to get the bounding boxes this is the shape I require in Windows ML. So, like channels_first_input I am expecting a channels_first_output such that if I want , I can enable else disable like channels_first_input.


As you said , Can I load the model in keras , do model.predict(data) and do transpose of that result array and add it back to model and then convert the yolo model into onnx ? Is it possible ? I just taught we can just add layers like convolution , max pooling but not transpose?
Is it correct?

**Comment by jiafatom** [comment]
@Rathna21 I just extract the first three layers, Conv2D, BatchNormalization, Activation Relu, and get the weights from original model. In this way I formulate a keras model and convert to onnx, the output result is correct. Then there should be something wrong in the remaining layers conversion. We need a sub model to debug this. 

It is not efficient I just extract the layers and set weights, because it is not easy to decompose the model from the whole model file.

You can also try yourself by looking at the examples in test_layers.py. there is assertTrue to compare results between keras and onnx. Since you have the source code, you can easily locate which layer is wrong by decomposing the model.

**Comment by Rathna21** [comment]
@jiafatom Thanks for the reply. Will try working on that and check. 

Do you need any sub model? If yes which layers do you require?

**Comment by jiafatom** [comment]
It is great if you can try testing because that is more flexible. You can also give the submodel where the last layer is the Concatenate. You have 4 Concatenate, then you can generate 4 sub models. Then it is good to compare each with onnx.

**Comment by Rathna21** [comment]
@jiafatom Correct as you said the model I shared with you has output shape (2,256,512,4). That is what even I want. Thats fine. The problem in that is my output values .

But I have another issue. I am working with yolo as well. I need to integrate the yolo model in Windows ML. YOLO in darknet is converted into keras using yad2k script. So, in keras it changes the input and output shape as (batch_size, height, width, channels) . But in windows ML its different. When I see yolo model in onnx model zoo, the output shape is (batch_size, channels , height, width) . Its the same in CORE ML as well. So, further to get the bounding boxes this is the shape I require in Windows ML. So, like channels_first_input I am expecting a channels_first_output such that if I want , I can enable else disable like channels_first_input.

As you said , Can I load the model in keras , do model.predict(data) and do transpose of that result array and add it back to model and then convert the yolo model into onnx ? Is it possible ? I just thought we can just add layers like convolution , max pooling but not transpose?
Is it correct?

**Comment by jiafatom** [comment]
@Rathna21 I don't have a definite answer for this last question. it is about data manipulation and out of converter scope....

**Comment by Rathna21** [comment]
@jiafatom , What do you mean sub model? Could you explain little more briefly?

**Comment by jiafatom** [comment]
@Rathna21 by sub model I mean that you can just extract the first few layers into one model, not all the layers.

**Comment by Rathna21** [comment]
@jiafatom Yea ok I will do that upload  the sub models.

And regarding the channels_first_output, you said you don't have any reason to implement it. But as the shape in Keras and other platforms is different , ie in Keras it is NHWC and most other platforms it is NCHW. Just to make onnx compatible with a framework with NCHW for input channels_first_input was enabled , could output also be done the same way so there will be no extra data manipulation at the end.

Ex: I want to work with YOLO in Windows ML. The problem was Windows ML would accept only NCHW to bind input to the model, whereas Keras was NHWC, I changed the input using keras2onnx to NCHW. I would expect to change my output shape as well. Even in darknet , the output of the yolo is in NCHW. 
So, I would like to work as such because in most of the frameworks it is NCHW expect Keras and tensorflow.
I would like to convert the darknet yolo model to convert to ONNX by using a intermediate framework which helps me support NCHW. But after a long research I see everyone using yad2k script to first convert into keras and then to any other framework. But only in keras it is changed to NHWC.

I tried removing the transpose layer in Keras. I face a lot of issues.

Is there a way you can help me in this? Of course most of them would be interested in NCHW output from this script as well.

**Comment by jiafatom** [comment]
I converted unet [here](https://github.com/zhixuhao/unet), and the converted onnx model gives me the correct result.

**Comment by Rathna21** [comment]
@jiafatom , Hello sorry for the delay in submodels.

I have created two subgraphs with concatenation layer at the end. Is this you were expecting?

https://drive.google.com/open?id=1BWss7g2wKFbO7eiWyzG2Ma7HXA0274x9
https://drive.google.com/open?id=12_vfwyzp752v-8rmq4VzSywhfV6sBkoj

Could you have a look into it?

**Comment by jiafatom** [comment]
@Rathna21 Yes, this is what we want. Both submodels gives correct result after conversion. Can we have more (end with the 3rd and 4th Concat)

**Comment by Rathna21** [comment]
@jiafatom ,

Yes, please find the remaining 2 models.

https://drive.google.com/open?id=1xhAY-1L-TcjJOyWFLhXBXu9PbR_wSJ4m

https://drive.google.com/open?id=1h0o7Z_qlS-PUth7Sh2jv4gCpCyeVMD7y

**Comment by jiafatom** [comment]
@Rathna21 Both models are correct, meaning that the submodel ending with all Concat is correct. So the remain difference is two sets of Conv2D/BatchNormalization/Acitvation, plus a last Conv2D. Can you give me two submodels, each end with Activation op (that is activation_24 and activation_25 in original unet_model.h5)?

**Comment by Rathna21** [comment]
@jiafatom , Ok will send the models. 

But, I have a question. Will have my weights changed , when I load the entire model and just extract some models from it?

Should I just load the weights instead?

**Comment by jiafatom** [comment]
@Rathna21 Not sure on this question. I randomly checked the weights for the submodels you send to me, looks same as original model. So please send the two models I mentioned, I can test it again.

**Comment by Rathna21** [comment]
@jiafatom ,

Please find the models you asked for!!

https://drive.google.com/open?id=1ikx1vfas2lKqmJxAhlFLbUdCPHqRI6KQ

https://drive.google.com/open?id=1WfY16pjmBP4RuNrmPgFb9s7WKVI3f7YD

**Comment by jiafatom** [comment]
@Rathna21 The results are correct. So the only difference is the last Conv2D. Can you do a test? Now you can extract any sub model, then you use this way to extract the whole model for me?

**Comment by Rathna21** [comment]
@jiafatom , You mean you want me to create a final sub model including last convolutional layer or test it by myself?

Dint get the exact thing you said?

**Comment by jiafatom** [comment]
@Rathna21 just checked in the [fix](https://github.com/onnx/keras-onnx/pull/69), and now your model is converted correctly. Please check, thanks.

**Comment by Rathna21** [comment]
@jiafatom 
It worked and values are perfect now.
Thank you so much for your assistance.
